### PR TITLE
Add memory spike test for out of order inserts

### PR DESCRIPTION
**Pull request by hjsuh18**
Added scripts and Travis cron job to test if out of order inserts cause memory spikes.

### CODE REVIEW

File: scripts/docker-run-memory-test.sh [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    done
+    exit 1
+}
+
+
+docker rm -f timescaledb-memory 2>/dev/null || true
+IMAGE_NAME=memory_test TAG_NAME=latest bash ${SCRIPT_DIR}/docker-build.sh
+
+docker_run timescaledb-memory memory_test:latest
+echo "Testing"
```
Comment (RobAtticus) [inline scripts/docker-run-memory-test.sh]: Move this down to above the last line, where it's actually doing the testing

File: scripts/docker-run-memory-test.sh [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+echo "Installing python3 and psutil"
+docker exec -u root -it timescaledb-memory /bin/bash -c "apk add --no-cache python3 && python3 -m ensurepip && pip3 install --upgrade pip && apk add --update build-base python3-dev py-psutil"
```
Comment (RobAtticus) [inline scripts/docker-run-memory-test.sh]: Pretty sure you can use the `docker_exec` bash function like the testing line. (`-u root` is the default so don't think it's necessary)

File: scripts/test_memory_spikes.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	for i in range (0, 5):
+		sum += p.memory_info().rss
+		time.sleep(3)
+	avg = sum / 5
+	print('Average memory consumption: ', bytes2human(avg))
+
+	upper_threshold = min(MEMORY_CAP, avg * THRESHOLD_RATIO)
+
+	# check if memory consumption goes above threshold until process terminates every 5 seconds
+	timeout = time.time() + 30 * 60
```
Comment (RobAtticus) [inline scripts/test_memory_spikes.py]: Travis got close to this, so maybe set to 45 or 50 mins?

File: scripts/test_memory_spikes.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	while True:
+		# insert finished
+		if not psutil.pid_exists(pid):
+			break
+		# prevent infinite loop
+		if time.time() > timeout:
+			print('Timed out on running inserts (over 30 mins)')
```
Comment (RobAtticus) [inline scripts/test_memory_spikes.py]: Update message if you change timeout above ^

File: scripts/test_memory_spikes.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+			p.kill()
+			sys.exit(4)
+
+		time.sleep(30)
+
+	print('No memory errors detected with out of order random inserts')
+	sys.exit(0) # success
+
+if __name__ == '__main__':
+	main()
```
Comment (RobAtticus) [inline scripts/test_memory_spikes.py]: add blank line at the end

File: scripts/docker-run-memory-test.sh [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+echo "Copying necessary from host to docker container"
+docker cp ${SCRIPT_DIR}/test_memory_spikes.py timescaledb-memory:/test_memory_spikes.py
+docker cp -a /tmp/tsdb-dev-tools timescaledb-memory:/tmp/tsdb-dev-tools
+
+docker_exec timescaledb-memory "python3 test_memory_spikes.py & cd /tmp/tsdb-dev-tools/ && sleep 5 && psql -U postgres -d postgres -h localhost -v ECHO=all -X -f sql/out_of_order_random_direct.sql"
```
Comment (RobAtticus) [inline scripts/docker-run-memory-test.sh]: add blank line at the end

File: .travis.yml [hunk 7]
```diff
@@ -46,6 +46,17 @@ jobs:
         # Versions <=9.6.1 is missing function `convert_tuples_by_name_map` needed by our code
         - PG_MAJOR=9.6 PG_MINOR_COMPILE=2 bash -x ./scripts/docker-run-abi-test.sh
 
+    - if: type = cron
+      stage: test
+      env:
+        - PG_VERSION=10.0 PG_GIT_TAG=REL_10_0
+        - secure: "CMZDBbHLIcSE01g+s8aqk72oV+cW8DKWcBMIlyfUPdk++PNPFIvafbMlMhvV+mWnJWh61W3zpTZfhYwuRk5JDKzsdE3PztMAPv81hTjaAypYQZbaaQeHK2rBfV47YF8osRPEQTQ9xVv9xl4Jlc5xvQoCYbHz2cM4vkfR05Z40qaDLAzR5wzX6cozDX/gSFLrfQ+8GhwBUj+5g0Eq36s2W2tcWBChlYqNylgeKDFh450lJsJQ3qvzzyU/2rVjEwyQ7V8XbJtDZvrKujRyl1XUrIaMDQgmPkrrX+/7/hgi6KVnks9OP6J480zmvVIitR27eE5g1+/791NlrPQJwhT81VnOci0d4Em4AbKal5Jh+ARKt6ovOYpbZgwU2XVRW9scv8E3OAe1TMvRiuyuP6by1tQdWgE4nqvCC2wbW08XQnNwaBtb+cQCgKlUv/zEG1wB0IgBQD//bRi9x2225o31WI8ql5uISDHbi6zat+DdzUef4+DeXRxQpI2V1GeoULZ0IhzSdiOgn3jPbItEXGZmG0qxdTfIs42EQWRPWXbYZeN6diXH735CSSJuFKkowjouhFWSQps3wfL2ToLPHt7lZU8ye9CmkpaGxi2la+W/p+rX5CtVR4m0ku/4xCPVKH7WFh4108QQ9EhImR9FPim+kty4TL/j1Scx0G9aJw/gHsE="
```
Comment (RobAtticus) [inline .travis.yml]: note to self: get ci-timescale's credentials to work

**Comment by RobAtticus** [comment]
Looks mostly good to me, just some minor nits

**Comment by cevian** [comment]
Why aren't these scripts going into tsdb-dev-tools? Seems unnecessary to have these scripts in main repo. @hjsuh18 @RobAtticus

**Comment by RobAtticus** [comment]
I can see arguments for both, hence the reason I proposed the question. Basically it would be nice if contributors can run the tests locally, in which case I think we might as well move the `.sql` file into the repo as well. If we prefer not to do that, then I agree the script files should go in the private `tsdb-dev-tools` repo.

File: scripts/test_memory_spikes.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
+# Python script to check if there are memory spikes when running out of 
+# order random inserts to TimescaleDB database
+import psutil
+import time
+import sys
+
+MEMORY_CAP = 300 * 1024 * 1024 # absolute memory cap
```
Comment (cevian) [inline scripts/test_memory_spikes.py]: Should these two variables be (optional) arguments to the program?

File: scripts/docker-run-memory-test.sh [hunk 9]
```diff
@@ -0,0 +1,83 @@
+#!/bin/bash
```
Comment (cevian) [inline scripts/docker-run-memory-test.sh]: I wonder if we should generalize this a bit and put everything before line 70 into some kind of bash util library. Do we do similar stuff in other /scripts/ files?

File: scripts/test_memory_spikes.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+			print('Killing postgres process')
+			p.kill()
+			sys.exit(4)
+
+		rss = p.memory_info().rss
+		print('Memory used by process ' + str(p.pid) + ': ' + bytes2human(rss))
+
+		# exit with error if memory above threshold
+		if rss > upper_threshold:
+			print('Memory consumption exceeded upper threshold')
```
Comment (cevian) [inline scripts/test_memory_spikes.py]: Did we ever test whether this actually fails the travis build?

File: scripts/test_memory_spikes.py (hunk 8 above)
Comment (RobAtticus) [reply scripts/test_memory_spikes.py]: We could. Don't feel strongly either way.

File: scripts/docker-run-memory-test.sh (hunk 9 above)
Comment (RobAtticus) [reply scripts/docker-run-memory-test.sh]: We should be wasn't sure throwing it into this PR was the right play; was thinking I would tackle afterward.

File: scripts/test_memory_spikes.py (hunk 10 above)
Comment (RobAtticus) [reply scripts/test_memory_spikes.py]: Yes, I ran it with a cap that is too small and Travis failed properly.

File: scripts/docker-run-memory-test.sh (hunk 9 above)
Comment (hjsuh18) [reply scripts/docker-run-memory-test.sh]: Yep, this part is simply copied from the `test_valgrind.sh` file. Same thing is used in `test_massif.sh`.

**Comment by RobAtticus** [comment]
@erimatnor - any thoughts on where we should store the scripts: here or in the private repo?

@cevian - comments addressed sufficiently?

File: scripts/docker-run-memory-test.sh [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+docker rm -f timescaledb-memory 2>/dev/null || true
+IMAGE_NAME=memory_test TAG_NAME=latest bash ${SCRIPT_DIR}/docker-build.sh
+
+docker_run timescaledb-memory memory_test:latest
+
+echo "Installing python3 and psutil"
+docker_exec timescaledb-memory "apk add --no-cache python3 && python3 -m ensurepip && pip3 install --upgrade pip && apk add --update build-base python3-dev py-psutil"
+
+echo "Copying necessary from host to docker container"
```
Comment (erimatnor) [inline scripts/docker-run-memory-test.sh]: necessary what?

File: .travis.yml [hunk 12]
```diff
@@ -46,6 +46,17 @@ jobs:
...
+        - secure: "S8WfPsrVGg21vZPxjrmy1X19T7zpHNgaSAoqYGzJdDZcNYuZbI87+tP3VQZo4MxjP3Ag46zTI5K+6aMjCqYqFGKufZdEaSpPSLmppFbXALgSp/rOkzp1M9HhGj0QVP7QcoPbva5Bjnj3zVi9CMAo6MwtnGAvzqBLOsdtnXTxEmKYyeNgwUVbQi2nzO38GeLgabaFL4XD3f3bmHE5jG6QAwZ/L7+jajsD1biH4sOiRLCRshtvermtYxh3SK6rML2kiwATABxTd1xDdMVRO/llmDBN373tHK6mufJyIhrwG9oTprSSDDIJlEbSPCkH0uvJBGcDvOAWf9tr4TGs/beawx3ELcXqR/4EisFqvMiVL4Vpt1gMBg9gue4Y0wpHZYoBOOef2gFtuyL7PjYe9koYJAJxg9DJ29DMxIjuCLTnbHN6yLa9425pdvxcuNEJ2K2zD0ZWoLSVueL0MxLhOZJsqlMkmmrFyI7y5cj1XvJF3vcsa5yK3S7sXroKUzeLu6QMh8hO3jD/2IvdQJ7Huy5uiJk2a+KUceLzruxLbwPF4b8SEoTquJ9NFGtagCG77lbIbLltLs2fUf+JnM8ipdpoMibVRFBOahP8NZwpL/NFzG3WLH3tDncj/c3fneMiNv3FUHLf6ufOVOayQiAEidszcMct2R20qIxggXKymrjZA3k="
+      before_script:
+        - git clone --progress --verbose https://$USR:$PASSWORD@bitbucket.org/440-labs/tsdb-dev-tools.git /tmp/tsdb-dev-tools
+      script:
+        - bash -x ./scripts/docker-run-memory-test.sh
```
Comment (erimatnor) [inline .travis.yml]: Security issue: Running bash with `-x` risks exposing secure variables. 

See here: https://docs.travis-ci.com/user/best-practices-security/

**Comment by hjsuh18** [comment]
I updated this PR addressing @erimatnor's comments. Is there a final decision on where to keep the scripts? Currently, I do a bit of both, having a python script in the main repo and also cloning `tsdb-dev-tools` to get the .sql file.

**Comment by RobAtticus** [comment]
You need to rebase and squash I believe, also here's a reworked commit desc:
```
Out of order, random inserts into TimescaleDB in the past
sometimes caused continuously increasing memory usage.
These checks are being added to ensure that there are no 
memory spike regressions via a cron Travis job.
```

**Comment by erimatnor** [comment]
@RobAtticus @hjsuh18 @cevian My take on where to put scripts and other testing stuff (apart from core tests) is that they should generally not be in the main repo. IMHO, we should avoid adding stuff to the main repo that few people will use or care about. I think that just increases the maintenance burden. Futher, the quality bar for what goes into this repo is higher than what we'd require for side-tools, which we mostly use for our internal testing.

I am not even sure  we need to run these tests off of the main repo? I imagine we could have another GitHub repo (e.g., `timescaledb-testing`) that just has the `.travis.yml` file for a bunch of cron jobs and then does a checkout of the `timescaledb` repo at runtime.

**Comment by RobAtticus** [comment]
Okay, we can move the scripts to the bitbucket repo.

I disagree on making a separate repo. I think it won't have as many eyes on it and therefore failing tests could go unnoticed for a while.

**Comment by hjsuh18** [comment]
OK I'll move the scripts to the bitbucket repo today. I'll initially assume that we are not going to create a separate repo for testing.

**Comment by hjsuh18** [comment]
@RobAtticus @cevian @erimatnor 
This PR has been updated, moved the python script into `tsdb-dev-tools`

**Pull request merged**